plugins/router/freebox

```diff
@@ -0,0 +1,316 @@
+#!/bin/bash
+# -*- sh -*-
+
+: << =cut
+
+=head1 NAME
+
+freebox - Plugin to monitor stats of a Freebox
+
+The ideal usage is also used as an updated warning limit.
+
+=head1 MAGIC MARKERS
+
+  #%# family=auto
+  #%# capabilities=autoconf
+
+=head1 CAVEAT
+
+Only tested on a Freebox v5 with an ADSL uplink.
+
+=head1 AUTHOR
+
+Olivier Mehani
+
+Copyright (C) 2019 Olivier Mehani <shtrom+munin@ssji.net>
+
+=head1 LICENSE
+
+SPDX-License-Identifier: GPL-3.0-or-later
+
+=cut
+
+# shellcheck disable=SC1090
+. "${MUNIN_LIBDIR}/plugins/plugin.sh"
+
+if [ "${MUNIN_DEBUG}" = 1 ]; then
+    set -x
+fi
+
+for CMD in curl iconv bc; do
+	if ! command -v "${CMD}" >/dev/null; then
+		echo "${CMD} not found" >&2
+		exit 1
+	fi
+done
+
+fbx_info_url="http://mafreebox.free.fr/pub/fbx_info.txt"
+
+fetch() {
+	# shellcheck disable=SC2154
+	curl -s "$@"
+}
+
+# XXX: testing
+# INFO="$( iconv -f latin1 -t utf8 fbx_info.txt)"
+INFO="$(fetch "${fbx_info_url}" | iconv -f latin1 -t utf8)"
+
+eval $(
```

Comment:
I am confused: what is the point of this `eval`?

Reply:
The `sed` just create a shell snippet that defines all the variables. The `eval` brings them into the context of the current script.

Reply:
Thank you for the clarification.
Personally I feel quite confident with writing shell scripts, but honestly I could only guess, what is happening here (due to the lack of understanding of the input text). This will be very hard for future contributors of this plugin :(
Maybe you can think of a more maintainable approach for this part?

Reply:
Sadly, no. The intput text (of which I added an example above) is very non machine-parsable, so it required a lot of `sed`-trickeries to get the relevant data out of it. I can add a comment explaining the generic idea, but yeah, essentially `sed` is used to extract the data and put it in the form of a number of variable assignments, for `bash` to evaluate.

Reply:
How about the following?
```shell
CRC_DOWN=$(echo "$INFO" | awk '/^\s+CRC\s/ { print $(NF-1); }')
CRC_UP=$(echo "$INFO" | awk '/^\s+CRC\s/ { print $NF; }')
```

Or maybe something even more readable?
```shell
get_line_column() {
    local key="$1"
    local field_index="$2"
    awk '/^\s+'"$key"'/ { print $('"$field_index"'); }'
}
...
CRC_DOWN=$(echo "$INFO" | get_line_column "CRC" "NF-1")
CRC_UP=$(echo "$INFO" | get_line_column "CRC" "NF")
```

Feel free to decide, whether such an approach could be suitable.

Reply:
Ok, this mostly works, but we lose validation of the fields. My `sed` script was ensuring that the extracted data was numeric, while the `awk` way gets whatever is in the column. This means that there is some bad data in case the output file doesn't have any.

For example, with the disconnected USB data:
```USB                    Non connecté```
we now get
```usb.value Non
usb_down.value   USB                    Non connecté 
```

And I still need the `eval $(sed)` to extract the uptime.

It seem to be slightly slower, too (by 100ms or so) [this is with a local dump of the file, rather than the network]:
```
[22:40:49] ~/src/munin-contrib/plugins/router$ for i in `seq 10`; do time MUNIN_LIBDIR=.  ./freebox &> /dev/null; done                                               130 ↵ ✹ router/freebox u+1 
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.14s user 0.08s system 104% cpu 0.214 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.19s user 0.05s system 105% cpu 0.228 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.05s system 105% cpu 0.208 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.07s system 104% cpu 0.229 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.14s user 0.07s system 104% cpu 0.204 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.05s system 105% cpu 0.209 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.07s system 104% cpu 0.224 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.17s user 0.09s system 104% cpu 0.251 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.16s user 0.09s system 103% cpu 0.240 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.16s user 0.06s system 104% cpu 0.201 total
[22:40:54] ~/src/munin-contrib/plugins/router$ git checkout -f HEAD^                                                                                                   
HEAD is now at afb3a6e4 fixup! fixup! [router/freebox] Plugin to monitor Freebox information
[22:40:59] ~/src/munin-contrib/plugins/router$ vim freebox                                                                                                   0s  (remotes/origin/router/freebox)
[22:41:06] ~/src/munin-contrib/plugins/router$ for i in `seq 10`; do time MUNIN_LIBDIR=.  ./freebox &> /dev/null; done                                     6s  ✹ (remotes/origin/router/freebox)
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.056 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.03s system 110% cpu 0.074 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.03s user 0.03s system 112% cpu 0.052 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.06s user 0.01s system 113% cpu 0.067 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.061 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.01s system 112% cpu 0.056 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.065 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.05s user 0.02s system 112% cpu 0.057 total
MUNIN_LIBDIR=. ./freebox &> /dev/null  0.04s user 0.02s system 112% cpu 0.059 total
```

Reply:
I pushed those changes. I'll let it run for a day or so to make sure I still get all the data, and let you know.

Reply:
Maybe the following is sufficient for numeric validation and proper handling of *unknown* values?
Bonus: only to be done once in the function :)

```shell
local result
result=$(awk '/^\s+'"$key"'/ { print $('"$field_index"'); }')
if echo "$result" | grep -q '^[0-9]\+$'; then
    echo "$result"
else
    echo "U"
fi
```

Regarding the uptime:
```shell
UPTIME_DAYS=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-5")
UPTIME_HOURS=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-3")
UPTIME_MINUTES=$(echo "$INFO" | get_line_column "Temps depuis la mise en route" "NF-1")
```

Maybe I am a bit obsessed about code readability :)

Reply:
Hum, I suppose that could work for the uptime. As for the validation, would there be a more direct way to match the regexp straight within `awk`? The `grep` should work, but it feels like a lot of overhead considering we already parsed the data out.

Reply:
Ok, I found a way. How about now?

Reply:
Yes, this should work easily by prepending the `print` call with something like `if ($('"$field_index"') =~ /^[0-9]+$/)`.

Reply:
This looks good to me!
Just one thing: maybe you need to add `^` in front of `/[.0-9]+/`? (otherwise `foo4` would match?)

Thank you for bearing with my wishes! I am happy to merge it: just say so.

Reply:
Ah, good catch on the `^...$` in the RE.

It's all fixed now, and squashed, so should be ready to merge :+1: .